packages/engine/src/datastore/batch/metaversion.rs

```diff
@@ -1,26 +1,230 @@
+//! other columns) to the agent state batch and sets the persisted version equal
+//! to its loaded version.
+//!
+//! * Behavior execution finishes and we reach the next step. To update context
+//! agent batches and handle any agent creation/removal messages, the main loop
+//! needs to have the latest version of agent state batches. Its loaded version, 1,
+//! is less than the persisted version, 2, so it loads the batch and sets the
+//! loaded version to the persisted version.
+//!
+//! * Behavior execution starts again. The main component of behavior execution
+//! writes behavior ids and sets the persisted version and its loaded version
+//! to 3.
+//!
+//! * The JavaScript component of behavior execution has loaded version 2,
+//! so it reloads the batch and sets its loaded version to 3. Then it runs
+//! behaviors, flushes data and sets the persisted version and its loaded
+//! version to 4.
+//!
+//! # Column-level metaversions
+//!
+//! Currently we only track batch-level metaversions, not column level metaversions.
+//! There would be two things to keep in mind with column-level metaversions:
+//!
+//! 1. The metaversions would need to be near the
+//! start of the shared memory segment and not move within it (or column/batch
+//! versions would be stored separately from memory versions). It helps that there's
+//! a fixed number of columns in a batch.
+//!
+//! 2. If some engine component is loading a column by just storing its location in
+//! memory and accessing it directly when necessary, then the loaded column can
+//! become invalid without being written to, e.g. if it's forced to move in memory
+//! due to another (e.g. string) column becoming larger.
+//! --> increment batch version, but not column versions when columns moved -->
+//! batch version = dynamic metadata version, column version = data version.
+//!
+//! If a simulation has e.g. a package `A` that only modifies column `a` and a package
+//! `B` that only modifies column `b`, it's better for performance not to flush changes
+//! more/earlier than necessary -- only when one needs to load the latest data or send
+//! it to a language runner.
 
-// Simple way for every component (language-runner + engine)
-// using the datastore to track whether it has to reload memory
-// or reload the recordbatch
+use crate::datastore::{storage::BufferChange, Error, Result};
 
+/// Simple way for every component (language runners + main loop)
+/// using the datastore to track whether it has to reload memory
+/// or reload the record batch.
 #[must_use]
-#[derive(Debug, Clone, Copy, Default)]
+#[derive(Debug, Clone, Copy, Default, PartialEq)]
 pub struct Metaversion {
     memory: u32,
     batch: u32,
 }
 
 impl Metaversion {
     // TODO: UNUSED: Needs triage
-    pub fn new(memory: u32, batch: u32) -> Metaversion {
-        Metaversion { memory, batch }
+    pub fn new(memory: u32, batch: u32) -> Result<Self> {
+        if !(batch >= memory) {
```

Comment:
I feel like `>=` is more readable here, because the invariant asserted elsewhere has `>=`, and the error message also has `>=`. But actually I think I have to remove this check entirely here.

Reply:
Addressed in #405